porcupine/plugins/git_status.py

```diff
@@ -0,0 +1,250 @@
+        bg = self.tree.tk.eval("ttk::style lookup Treeview -background")
+        gray = utils.mix_colors(fg, bg, 0.5)
+
+        if sum(self.tree.winfo_rgb(fg)) > 3 * 0x7FFF:
+            # bright foreground color
+            green = "#00ff00"
+            orange = "#ff6e00"
+        else:
+            green = "#007f00"
+            orange = "#e66300"
+
+        self.tree.tag_configure("git_mergeconflict", foreground=orange)
+        self.tree.tag_configure("git_modified", foreground="red")
+        self.tree.tag_configure("git_added", foreground=green)
+        self.tree.tag_configure("git_untracked", foreground="red4")
+        self.tree.tag_configure("git_ignored", foreground=gray)
+
+    def start_status_coloring_for_all_projects(self, junk_event: object) -> None:
+        old_project_ids = set(self.project_specific_colorers.keys())
+        new_project_ids = set(self.tree.get_children())
+        for removed in old_project_ids - new_project_ids:
+            del self.project_specific_colorers[removed]
+        for added in new_project_ids - old_project_ids:
+            self.project_specific_colorers[added] = ProjectColorer(self.tree, added)
+
+        for colorer in self.project_specific_colorers.values():
+            colorer.start_running_git_status()
+
+    def update_git_tags_for_item(self, item_id: str) -> None:
+        project_id = self.tree.find_project_id(item_id)
+        self.project_specific_colorers[project_id].update_colors_now_or_later(item_id)
+
+
+# There's no way to say "when this item is selected, show a green selection".
+# But there is a way to say "when any item is selected, show a green selection".
+# We just have to configure that whenever a different item is selected, lol
+def update_tree_selection_color(tree: DirectoryTree, event: object = None) -> None:
+    try:
+        [selected_id] = tree.selection()
+    except ValueError:  # nothing selected
+        git_tags = []
+    else:
+        git_tags = [tag for tag in tree.item(selected_id, "tags") if tag.startswith("git_")]
+
+    if git_tags:
+        [tag] = git_tags
+        color = tree.tag_configure(tag, "foreground")
+        tree.tk.call(
+            "ttk::style",
+            "map",
+            "DirectoryTreeGitStatus.Treeview",
+            "-foreground",
+            ["selected", color],
+        )
+    else:
+        # use default colors
+        tree.tk.eval("ttk::style map DirectoryTreeGitStatus.Treeview -foreground {}")
+
+
+def sorting_key(tree: DirectoryTree, item_id: str) -> Any:
+    [git_tag] = [t for t in tree.item(item_id, "tags") if t.startswith("git_")] or [None]
+    return [
+        "git_added",
+        "git_modified",
+        "git_mergeconflict",
+        None,
+        "git_untracked",
+        "git_ignored",
+    ].index(git_tag)
+
+
+def setup() -> None:
+    for tree in utils.get_children_recursively(get_paned_window()):
+        if isinstance(tree, DirectoryTree):
```

Comment:
This logic appears in many places now, including tests. Maybe there should be a function that finds and returns the tree from the currently running Porcupine?